### PR TITLE
fix: Range suffixes are not Rust RangeTo

### LINKED ISSUE #87: Mapping of HTTP ranges to std::ops::Bound based ranges is faulty
tl;dr: `bytes=-x` is being mapped to `..x` by this crate, where http headers mean the last `x` bytes, and rust `BoundRanges` mean the first `x` bytes.

#### HTTP Ranges
For HTTP ranges, there are four options:
 1. Not defining a range. This means everything is included
 2. `Range: <unit>=<range-start>-`, where only the range beginning is given. This means everything from `start` is included.
 3. `Range: <unit>=<range-start>-<range-end>`, where both the start and end is given. This means everything between `start` and `end` are included
 4. `Range: <unit>=-<suffix-length>`, where just a suffix is given. This means only the tail, the last `suffix` bytes are returned.

#### Rust RangeBounds
Rust's `RangeBounds` *also* have (kinda) four options:
 1. `..`, which includes everything.
 2. `start..`, which includes everything except the first `start` bytes.
 3. `start..end` (or `start..=(end - 1)`), which includes everything from `start` to `end`, excluding `end`.
 4. `..end` (or `..=(end - 1)`), which includes everything from 0 to `end`, excluding `end`.

### Current Mapping between them
Currently, this mapping is just done one-to-one based on their syntax, not based on their definition. This means that `Range: bytes=x-y` is mapped to `x..y`, which is fine if both `x` and `y` are defined. Only including `x` is also fine, as they'll still mean the same. But, if only have `y`, the mapping becomes wrong.

There is no correct mapping for these cases, as "give me the last `n` bytes" is not something you can represent in Rust's `RangeBounds`. When the length we're dealing with is known, conversion is possible, but `RangeBounds` are probably not the right type in that case for storing the header information, as converting back isn't possible.

#### Require length before returning ranges
Given the length, converting from HTTP Ranges is possible. This would mean that the signature of `headers::Range::iter` would need to be changed, to include the length of the content being returned.

#### Don't use RangeBounds
Alternatively to giving the length, we could use a separate type here, and have a conversion method (which takes the length again) to map it to `RangeBounds`.

### PR DESCRIPTION
**Pull request by seanmonstar**
An HTTP Range of `bytes=-100` means a suffix, the last 100 bytes. This was wrongly parsed as the Rust range `..100`, which means the first 100 bytes.

This has been fixed, but doing so required change `Range::iter` to accept a length argument, to determine if the ranges are satisfiable.

Closes #87 

BREAKING CHANGE: Change `.iter()` calls to `.satisfiable_ranges(len)`.
  Also, the `Range::bytes()` constructor will now return an error if
  pass a `RangeTo` (e.g. `Range::bytes(..100)`).

### CODE REVIEW

**Comment by seanmonstar** [comment]
This waited so long because it was a breaking change, but can now be included with the `http` upgrade. This is a different take from #93, which doesn't expose as many new types.

It's a very conservative change, while allowing something much more comprehensive from eventually being added.

**Review by GlenDC — APPROVED** [review]

File: src/common/range.rs [hunk 1]
```diff
@@ -416,3 +439,10 @@ fn test_byte_range_spec_to_satisfiable_range() {
 bench_header!(bytes_multi, Range, { vec![b"bytes=1-1001,2001-3001,10001-".to_vec()]});
 bench_header!(custom_unit, Range, { vec![b"other=0-100000".to_vec()]});
 */
+
+#[test]
+fn test_to_satisfiable_range_suffix() {
+    let range = super::test_decode::<Range>(&["bytes=-100"]).unwrap();
+    let bounds = range.satisfiable_ranges(350).next().unwrap();
+    assert_eq!(bounds, (Bound::Included(250), Bound::Unbounded));
+}
```
Comment (rakshith-ravi) [inline src/common/range.rs:448]: Should we also include this?

```suggestion
}

#[test]
fn test_to_unsatisfiable_range_suffix() {
    let range = super::test_decode::<Range>(&["bytes=-350"]).unwrap();
    let bounds = range.satisfiable_ranges(100).next();
    assert_eq!(bounds, None);
}
```

**Comment by rakshith-ravi** [comment]
Idk if you're big on tests, but if we're testing satisfiable range, then the opposite should also be tested, shouldn't it?

File: src/common/range.rs [hunk 2]
```diff
@@ -48,29 +48,52 @@ impl Range {
...
-    /// Iterate the range sets as a tuple of bounds.
-    pub fn iter<'a>(&'a self) -> impl Iterator<Item = (Bound<u64>, Bound<u64>)> + 'a {
+    /// Iterate the range sets as a tuple of bounds, if valid with length.
+    ///
+    /// The length of the content is passed as an argument, and all ranges
+    /// that can be satisfied will be iterated.
+    pub fn satisfiable_ranges<'a>(
+        &'a self,
+        len: u64,
+    ) -> impl Iterator<Item = (Bound<u64>, Bound<u64>)> + 'a {
```
Comment (DDtKey) [inline src/common/range.rs:72]: Hi @seanmonstar 
Thanks for these changes,  but it looks like it still works not as expected 

We can pass the `len` of the content, but for `unbounded` or `included` > `len` we don't use `min()`, like for example it was implemented in https://github.com/hyperium/headers/pull/93

And in general would be nice to have more validation, like `from` < `to`?

**Pull request merged**